File: operator-api/src/routes/macros.js
import { Hono } from 'hono'
import { uid } from '../utils/ids.js'
import { execCapture } from '../utils/exec.js'
import 'dotenv/config'

const macros = new Map() // id -> {name, steps}
const display = process.env.DISPLAY || ':0'
const XDOTOOL = process.env.XDOTOOL_BIN || '/usr/bin/xdotool'

export const macrosRouter = new Hono()

async function runXdotool(args) {
  const res = await execCapture(XDOTOOL, args, { env: { ...process.env, DISPLAY: display } })
  if (res.code !== 0) throw new Error(res.stderr.toString('utf8') || 'xdotool error')
  return res
}

macrosRouter.post('/macros/create', async (c) => {
  const body = await c.req.json()
  if (!body?.name || !Array.isArray(body.steps)) return c.json({ message: 'Bad Request' }, 400)
  const id = uid()
  macros.set(id, { macro_id: id, name: body.name, steps: body.steps })
  return c.json({ macro_id: id })
})

macrosRouter.post('/macros/run', async (c) => {
  const body = await c.req.json()
  const item = [...macros.values()].find((m) => m.macro_id === body.macro_id)
  if (!item) return c.json({ message: 'Not Found' }, 404)
  const run_id = uid()
    ; (async () => {
      for (const step of item.steps) {
        if (step.action === 'keyboard.type' && step.text) await runXdotool(['type', '--clearmodifiers', '--', step.text])
        else if (step.action === 'keyboard.key' && step.key) await runXdotool(['key', step.key])
        else if (step.action === 'sleep' && step.ms) await runXdotool(['sleep', String(step.ms / 1000)])
      }
    })()
  return c.json({ started: true, run_id })
})

macrosRouter.get('/macros/list', async (c) => {
  return c.json({ items: [...macros.values()].map((m) => ({ macro_id: m.macro_id, name: m.name, steps_count: m.steps.length })) })
})

macrosRouter.delete('/macros/:macro_id', async (c) => {
  const id = c.req.param('macro_id')
  if (!macros.has(id)) return c.json({ message: 'Not Found' }, 404)
  macros.delete(id)
  return c.json({ ok: true })
})

File: server-sigma/internal/routes/macros/macros.go
package macros

import (
	"encoding/json"
	"net/http"
	"os"
	"sync"

	"github.com/go-chi/chi/v5"

	"kernel-operator-api/internal/utils"
)

type Macro struct {
	MacroID string         `json:"macro_id"`
	Name    string         `json:"name"`
	Steps   []map[string]any `json:"steps"`
}

var (
	mu     sync.Mutex
	macros = map[string]*Macro{}
)

var (
	display = func() string { d := os.Getenv("DISPLAY"); if d == "" { d = ":0" }; return d }()
	xdotool = func() string { x := os.Getenv("XDOTOOL_BIN"); if x == "" { x = "/usr/bin/xdotool" }; return x }()
)

func runXdotool(args ...string) error {
	res := utils.Capture(xdotool, args, map[string]string{"DISPLAY": display}, "", 0)
	if res.Code != 0 {
		return &xdErr{stderr: string(res.Stderr)}
	}
	return nil
}
type xdErr struct{ stderr string }
func (e *xdErr) Error() string { return e.stderr }

func Router() *chi.Mux {
	r := chi.NewRouter()

	r.Post("/macros/create", func(w http.ResponseWriter, r *http.Request) {
		var b struct {
			Name  string           `json:"name"`
			Steps []map[string]any `json:"steps"`
		}
		_ = json.NewDecoder(r.Body).Decode(&b)
		if b.Name == "" || len(b.Steps) == 0 {
			http.Error(w, `{"message":"Bad Request"}`, 400); return
		}
		id := utils.UID()
		m := &Macro{MacroID: id, Name: b.Name, Steps: b.Steps}
		mu.Lock(); macros[id] = m; mu.Unlock()
		json.NewEncoder(w).Encode(map[string]any{"macro_id": id})
	})

	r.Post("/macros/run", func(w http.ResponseWriter, r *http.Request) {
		var b struct{ MacroID string `json:"macro_id"` }
		_ = json.NewDecoder(r.Body).Decode(&b)
		mu.Lock(); m := macros[b.MacroID]; mu.Unlock()
		if m == nil {
			http.Error(w, `{"message":"Not Found"}`, 404); return
		}
		runID := utils.UID()
		go func() {
			for _, step := range m.Steps {
				switch step["action"] {
				case "keyboard.type":
					if t, ok := step["text"].(string); ok {
						_ = runXdotool("type", "--clearmodifiers", "--", t)
					}
				case "keyboard.key":
					if k, ok := step["key"].(string); ok {
						_ = runXdotool("key", k)
					}
				case "sleep":
					if ms, ok := step["ms"].(float64); ok {
						runXdotool("sleep", formatFloat(ms/1000.0))
					}
				}
			}
		}()
		json.NewEncoder(w).Encode(map[string]any{"started": true, "run_id": runID})
	})

	r.Get("/macros/list", func(w http.ResponseWriter, r *http.Request) {
		type item struct {
			MacroID    string `json:"macro_id"`
			Name       string `json:"name"`
			StepsCount int    `json:"steps_count"`
		}
		mu.Lock()
		var items []item
		for _, m := range macros {
			items = append(items, item{MacroID: m.MacroID, Name: m.Name, StepsCount: len(m.Steps)})
		}
		mu.Unlock()
		json.NewEncoder(w).Encode(map[string]any{"items": items})
	})

	r.Delete("/macros/{id}", func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		mu.Lock(); _, ok := macros[id]; delete(macros, id); mu.Unlock()
		if !ok {
			http.Error(w, `{"message":"Not Found"}`, 404); return
		}
		json.NewEncoder(w).Encode(map[string]any{"ok": true})
	})

	return r
}

func formatFloat(f float64) string {
	s := strconvIt(int(f * 1000))
	// already seconds with 3 decimals
	return s[:len(s)-3] + "." + s[len(s)-3:]
}
func strconvIt(i int) string { return itoa(i) }
func itoa(i int) string { return fmtInt(i) }

// tiny int -> string
func fmtInt(i int) string {
	if i == 0 { return "0" }
	sign := ""
	if i < 0 { sign = "-"; i = -i }
	d := []byte{}
	for i > 0 {
		d = append([]byte{byte('0'+i%10)}, d...)
		i/=10
	}
	return sign+string(d)
}